pallets/domains/src/lib.rs

```diff
@@ -0,0 +1,517 @@
+            let domain_meta = DomainMeta::new(
+                full_domain.clone(),
+                owner.clone(),
+                content,
+                expires_at,
+                deposit,
+            );
+
+            if let ReserveDeposit::Yes = reserve_deposit {
+                <T as Config>::Currency::reserve(&owner, deposit)?;
+            }
+
+            // TODO: withdraw balance
+
+            RegisteredDomains::<T>::insert(domain_lc.clone(), domain_meta);
+            DomainsByOwner::<T>::mutate(
+                &owner, |domains| {
+                    domains.try_push(domain_lc.clone()).expect("qed; too many domains per account")
+                }
+            );
+
+            Self::deposit_event(Event::DomainRegistered(owner, full_domain));
+            Ok(())
+        }
+
+        /// Throws an error if domain contains invalid character.
+        fn ensure_domain_contains_valid_chars(domain: &[u8], error: Error<T>) -> DispatchResult {
+            let is_char_alphanumeric = |c: &&u8| (**c).is_ascii_alphanumeric();
+
+            let first_char_alphanumeric = domain.first().filter(is_char_alphanumeric).is_some();
+            let last_char_alphanumeric = domain.last().filter(is_char_alphanumeric).is_some();
+
+            let mut prev_char_hyphen = false;
+            let domain_correct = domain.iter().all(|c| {
+                let curr_char_hyphen = *c == b'-';
+
+                if prev_char_hyphen && curr_char_hyphen {
+                    return false;
+                }
+
+                prev_char_hyphen = curr_char_hyphen;
+                c.is_ascii_alphanumeric() || curr_char_hyphen
+            });
+
+            ensure!(first_char_alphanumeric && last_char_alphanumeric && domain_correct, error);
+
+            Ok(())
+        }
+
+        /// The domain must match the recommended IETF conventions:
+        /// https://datatracker.ietf.org/doc/html/rfc1035#section-2.3.1
+        ///
+        /// The domains must start with a letter, end with a letter or digit,
+        /// and have as interior characters only letters, digits, and/or hyphens.
+        /// There are also some restrictions on the length:
+        /// Domains length must be between 3 and 63 characters.
+        pub fn ensure_valid_domain(domain: &[u8]) -> DispatchResult {
+            let mut split = domain.split(|c| *c == b'.');
+            let dots = split.clone().count().saturating_div(2);
+
+            ensure!(dots <= 1, Error::<T>::LowerLevelDomainsNotAllowed);
+            ensure!(!dots.is_zero(), Error::<T>::TopLevelDomainNotSpecified);
+
+            let domain = split.next().unwrap();
+            ensure!(split.next().unwrap() == TOP_LEVEL_DOMAIN, Error::<T>::TopLevelDomainNotAllowed);
+
+            // No need to check max length, because we use BoundedVec as input value.
+            ensure!(
+                domain.len() >= T::MinDomainLength::get() as usize,
+                Error::<T>::DomainNameIsTooShort,
+            );
+
```

Comment:
Also need to check that domain length is not too long.

Reply:
No need, because we use `BoundedVec` which will not allow us to even pass such value as an extrinsic argument.